### PR TITLE
feat(Scripts): support angularJS apps

### PR DESCRIPTION
**Pull request by jsomsanith**
**What is the problem this PR is trying to solve?**
No support of angular

**What is the chosen solution to this problem?**
* add a `test:ng` goal to run tests via karma, using jasmine
* add angular eslint rules with eslint plugin angular
* talend-scripts.json > angular-legacy option
  * add `ng-annotate`
  * add `ngtemplate-loader`

Other changes
* start does not open the app in a new tab anymore. This was annoying for some apps, that must serve it via a gateway. To open it again, just pass the option `talend-scripts start --open`
* add cache-loader to increase build speed 
* replace `mini-css-extract-plugin` by `style-loader` in dev mode
* bundle split vendor/app/polyfill
* configure optimization for prod bundle
* add a mock to session storage and window location for jest

**Please check if the PR fulfills these requirements**

* [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [ ] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/scripts/package.json [hunk 1]
```diff
@@ -45,33 +47,53 @@
...
     "uglifyjs-webpack-plugin": "1.2.7",
     "url-loader": "^1.1.2",
     "webpack": "^4.19.0",
     "webpack-bundle-analyzer": "^3.0.4",
     "webpack-cli": "^3.1.0",
     "webpack-dev-server": "^3.1.11",
     "whatwg-fetch": "^2.0.3",
-    "which": "^1.3.0"
+    "which": "^1.3.0",
+    "yargs": "^13.2.1"
```
Comment (frassinier) [inline packages/scripts/package.json]: not usefull, please remove

File: output/components.sasslint.txt [hunk 2]
```diff
@@ -10,12 +10,18 @@ src/Drawer/Drawer.scss
 src/FilterBar/FilterBar.scss
   48:6  warning  Vendor prefixes should not be used  no-vendor-prefixes
 
+src/List/Toolbar/SelectSortBy/SelectSortBy.scss
```
Comment (frassinier) [inline output/components.sasslint.txt]: wat?

File: packages/scripts/webapp/preset/README.md [hunk 3]
```diff
@@ -35,6 +35,7 @@ Your folder hierarchy should follow
 ```json
 {
   "preset": "talend",
+  "angular-legacy": true,
```
Comment (frassinier) [inline packages/scripts/webapp/preset/README.md]: `angular`instead of `angular-legacy` is enough

File: packages/scripts/webapp/preset/README.md [hunk 4]
```diff
@@ -281,6 +283,26 @@ To use the eslint configuration in your IDE
 
 3. Configure your IDE plugin to enable eslint with your root eslintrc configuration.
 
+## Angular tests via Karma
+
+To run angular tests via karma, you need to create a `spec.bundle.js` in your project root.
+It must include everything Karma needs to load to run your tests.
```
Comment (frassinier) [inline packages/scripts/webapp/preset/README.md]:
```suggestion
It must include everything that Karma needs to load to run your tests.
```

File: packages/scripts/webapp/preset/config/webpack.config.dev.js [hunk 5]
```diff
@@ -1,26 +1,16 @@
...
+		aggregateTimeout: 300,
+		poll: 1000,
+	},
+	devServer: {
+		port: 3000,
+		stats: 'errors-only',
+		historyApiFallback: true,
+		contentBase: path.join(process.cwd(), '/dist'),
+	},
+	devtool: 'eval',
```
Comment (frassinier) [inline packages/scripts/webapp/preset/config/webpack.config.dev.js]: why is it now set to eval?

File: packages/scripts/webapp/preset/config/webpack.config.js [hunk 6]
```diff
@@ -37,19 +34,19 @@ function getCommonStyleLoaders(enableModules, mode) {
...
 	}
 	return [
-		{ loader: MiniCssExtractPlugin.loader },
+		{ loader: mode === 'development' ? 'style-loader' : MiniCssExtractPlugin.loader },
 		{ loader: 'css-loader', options: cssOptions },
 		{
 			loader: 'postcss-loader',
 			options: {
-				sourceMap: mode !== 'production',
+				sourceMap: mode === 'development',
```
Comment (frassinier) [inline packages/scripts/webapp/preset/config/webpack.config.js]: DRY, you can use a variable 😎

File: packages/scripts/webapp/preset/config/webpack.config.js [hunk 7]
```diff
@@ -121,11 +118,80 @@ function getVersions() {
...
+	if (angularLegacy) {
+		loaders.unshift({ loader: 'ngtemplate-loader' });
+	}
+
+	return loaders;
+}
+
+function getCopyConfig(userCopyConfig = []) {
+	const config = [...userCopyConfig];
+	const assetsOverridden = config.some(({ from }) => from === 'src/assets');
```
Comment (frassinier) [inline packages/scripts/webapp/preset/config/webpack.config.js]: if `from` is equals to `to`, you can pass an array https://github.com/webpack-contrib/copy-webpack-plugin#from

File: packages/scripts/webapp/preset/config/webpack.config.js [hunk 8]
```diff
@@ -121,11 +118,80 @@ function getVersions() {
...
+	const { theme } = userSassData;
+
+	const sassData = getSassData(userSassData);
+	const indexTemplatePath = `${process.cwd()}/src/app/index.html`;
 
 	return {
-		entry: ['@babel/polyfill', 'whatwg-fetch', `${process.cwd()}/src/app/index.js`],
+		entry: {
+			polyfills: ['@babel/polyfill', 'whatwg-fetch'],
+			style: [
```
Comment (frassinier) [inline packages/scripts/webapp/preset/config/webpack.config.js]: `theme` instead?

File: packages/scripts/webapp/preset/config/webpack.config.js [hunk 9]
```diff
@@ -196,22 +264,27 @@ module.exports = ({ getUserConfig, mode }) => {
 				BUILD_TIMESTAMP: Date.now(),
 				TALEND_APP_INFO: JSON.stringify(getVersions()),
 			}),
-			new MiniCssExtractPlugin({
-				filename: '[name]-[hash].css',
-			}),
+			mode === 'development' && new webpack.HotModuleReplacementPlugin({}),
```
Comment (frassinier) [inline packages/scripts/webapp/preset/config/webpack.config.js]: do we need empty option?

File: packages/scripts/webapp/preset/config/webpack.config.dev.js [hunk 10]
```diff
@@ -1,26 +1,16 @@
...
+	mode: 'development',
+	watchOptions: {
+		aggregateTimeout: 300,
+		poll: 1000,
+	},
+	devServer: {
+		port: 3000,
+		stats: 'errors-only',
+		historyApiFallback: true,
+		contentBase: path.join(process.cwd(), '/dist'),
```
Comment (frassinier) [inline packages/scripts/webapp/preset/config/webpack.config.dev.js]: We should add `compress: true` and `hot: true`

File: packages/scripts/webapp/scripts/start.js [hunk 11]
```diff
@@ -10,7 +10,6 @@ module.exports = function start(env, _, options) {
 			'--inline',
 			'--config',
 			hereRelative(__dirname, '../config/webpack.config.js'),
-			'--open',
 			'--progress',
 			'--display-error-details',
```
Comment (frassinier) [inline packages/scripts/webapp/scripts/start.js]: we should avoid mix CLI args and config file, it's a mess.

File: packages/scripts/package.json (hunk 1 above)
Comment (jsomsanith) [reply packages/scripts/package.json]: In fact it is used in `karma.conf.js` to set the single run arg

File: output/components.sasslint.txt (hunk 2 above)
Comment (jsomsanith) [reply output/components.sasslint.txt]: 0_0

File: packages/scripts/webapp/preset/README.md (hunk 3 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/README.md]: Changed

File: packages/scripts/webapp/preset/config/webpack.config.dev.js (hunk 5 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/config/webpack.config.dev.js]: Do you prefer cheap-module-eval-source-map ? (https://webpack.js.org/configuration/devtool/ )

The old one was source map that is not dev friendly in build/rebuild time

File: packages/scripts/webapp/preset/config/webpack.config.js (hunk 6 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/config/webpack.config.js]: It was more to not having mode/isProdMode/isDevMode depending on where you are.
Here the args are consistent.
:/

File: packages/scripts/webapp/preset/config/webpack.config.js (hunk 7 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/config/webpack.config.js]: Changed to check if we have an object or not

File: packages/scripts/webapp/preset/config/webpack.config.dev.js (hunk 5 above)
Comment (frassinier) [reply packages/scripts/webapp/preset/config/webpack.config.dev.js]: `cheap-module-source-map` for dev and `source-map` for th rest, yes.

File: packages/scripts/package.json (hunk 1 above)
Comment (frassinier) [reply packages/scripts/package.json]: flag could not be passed from @talend/scripts...

File: packages/scripts/webapp/scripts/start.js (hunk 11 above)
Comment (jsomsanith) [reply packages/scripts/webapp/scripts/start.js]: I remove `inline` (it's the default)
`progress` is cli only
`display-error-details` ... I don't find any documentation on that :/

File: packages/scripts/webapp/preset/config/webpack.config.js (hunk 8 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/config/webpack.config.js]: Done

File: packages/scripts/webapp/preset/config/webpack.config.js (hunk 9 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/config/webpack.config.js]: Removed

File: packages/scripts/webapp/preset/config/webpack.config.dev.js (hunk 10 above)
Comment (jsomsanith) [reply packages/scripts/webapp/preset/config/webpack.config.dev.js]: Added

**Pull request merged**